server/src/main/java/org/opensearch/gateway/PrimaryShardAllocator.java

```diff
@@ -393,6 +403,11 @@ protected static NodeShardsResult buildNodeShardsResult(
             comparator = NO_STORE_EXCEPTION_FIRST_COMPARATOR.thenComparing(PRIMARY_FIRST_COMPARATOR);
         }
 
+        // If index has segrep enabled, then use replication checkpoint info to order the replicas
+        if (indexSettings != null && indexSettings.isSegRepEnabled()) {
+            comparator = comparator.thenComparing(HIGHEST_REPLICATION_FIRST_CHECKPOINT_COMPARATOR);
```

Comment:
Is HIGHEST_REPLICATION_FIRST_CHECKPOINT_COMPARATOR a no-op in the document replication scenario? If so, can we consider not branching based on whether seg rep is enabled and always include the comparator?

Reply:
Since, the change here adds ReplicationCheckpoint to `NodeGatewayStartedShards` without feature flag, enabling this comparator for doc_rep will indeed order replicas based on replication checkpoints. This is good as it chooses the furthest ahead replica. But, as this change is not explicitly seeked for doc_rep; added it behind feature flag.

Reply:
For doc based wouldn't checkpoints be returned as null/empty-values based on ser/de and compare a no-op? I think that is what @andrross is meaning to suggest as well

Reply:
Thanks @andrross @Bukhtawar for the review. Yes, this comparator will be no-op for doc-rep based indices as [getLatestReplicationCheckpoint](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L1402) method is not behind a feature flag. So, this will result in non-null `ReplicationCheckpoint` from the data nodes.

Reply:
>I see null is returned from TransportNodesListGatewayStartedShards, I think we could update getLatestReplicationCheckpoint to return null instead of empty with the FF disabled - I think we are missing that check.

Created PR for adding FF check https://github.com/opensearch-project/OpenSearch/pull/4163

Reply:
Removed the `if` condition as it is NO-OP for doc rep enabled indices.